### PR TITLE
Extract computations from ReflectometryMomentumTransfer to a separate algorithm

### LINKED ISSUE #23622: ReflectometryILL: algorithm for beam statistics
`ReflectometryMomentumTransfer` calculates a number of statistical quantities related to the beam divergence which are needed for the Q resolution calculation within the algorithm. Some of these quantities (*sample waviness* and *slit 1 angular spread* at least) are needed also to determine if the sample is bent in `ReflectometrySumInQ`.

To avoid calculating these quantities in both `ReflectometryMomentumTransfer` and `ReflectometrySumInQ`, the calculation could be offloaded to a new algorithm to be called in `ReflectometryILLSumForeground`. The results could be saved in the workspace's sample logs, or in PropertyManager. This would also allow the removal of the confusing *DirectBeamWorkspace* and *ReflectedBeamWorkspace* input properties from `ReflectometryILLConvertToQ`.

### PR DESCRIPTION
**Pull request by soininen**
**Description of work.**

Extracted the calculation of certain statistical quantities related to the beam divergence etc. from  `ReflectometryMomentumTransfer` to a new algorithm, `ReflectometryBeamStatistics` simplifying the ILL's reflectometry reduction workflow. The said quantities are also needed in `ReflectometrySumInQ` so we avoid calculating things twice.

**Report to:** [nobody].

**To test:**

Test data is available in the unit test data directory. Make sure you have the data search directories set up correctly.

Run the test script:
```python
ReflectometryILLPreprocess(
    'ILL/D17/317369',
    OutputWorkspace='direct',
    ForegroundHalfWidth=[2, 3])

ReflectometryILLPreprocess(
    'ILL/D17/317370',
    OutputWorkspace='reflected',
    ForegroundHalfWidth=[3])

def foreground(ws):
    fgd = list()
    run = mtd[ws].run()
    fgd.append(run.getProperty('foreground.first_workspace_index').value)
    fgd.append(run.getProperty('foreground.centre_workspace_index').value)
    fgd.append(run.getProperty('foreground.last_workspace_index').value)
    return fgd

ReflectometryBeamStatistics(
    ReflectedBeamWorkspace='reflected',
    ReflectedForeground=foreground('reflected'),
    DirectBeamWorkspace='direct',
    DirectForeground=foreground('direct'),
    PixelSize=0.001195,
    DetectorResolution=0.0022,
    FirstSlitName='slit2',
    FirstSlitSizeSampleLog='VirtualSlitAxis.s2w_actual_width',
    SecondSlitName='slit3',
    SecondSlitSizeSampleLog='VirtualSlitAxis.s3w_actual_width')
```

Check that the `reflected` workspace now has the following sample log entries: `beam_stats.beam_rms_variation`, `beam_stats.bent_sample`, `beam_stats.first_slit_angular_spread`, `beam_stats.incident_angular_spread`, `beam_stats.sample_waviness`, `beam_stats.second_slit_angular_spread`.

Check that the `direct` workspace now has the `beam_stats.beam_rms_variation` sample log entry.

To test that `ReflectometryMomentumTransfer` still works fine and utilizes the log entries added by `ReflectometryBeamStatistics`, run the following script (uses the workspaces from the script above):
```python
run = mtd['reflected'].run()
first = run.getProperty('foreground.first_workspace_index').value
last = run.getProperty('foreground.last_workspace_index').value
SumSpectra(
    InputWorkspace='reflected',
    OutputWorkspace='reflected-foreground',
    StartWorkspaceIndex=first,
    EndWorkspaceIndex=last)

ReflectometryMomentumTransfer(
    'reflected-foreground',
    OutputWorkspace='R_Q',
    ReflectedForeground=[first, last],
    PixelSize=0.001195,
    DetectorResolution=0.0022,
    ChopperSpeed=1000.,
    ChopperOpening=3.,
    ChopperRadius=0.3,
    ChopperPairDistance=0.4,
    FirstSlitName='slit2',
    FirstSlitSizeSampleLog='VirtualSlitAxis.s2w_actual_width',
    SecondSlitName='slit3',
    SecondSlitSizeSampleLog='VirtualSlitAxis.s3w_actual_width',
    TOFChannelWidth=57.)
```

The script should execute without issues.

Fixes #23622.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/ReflectometryBeamStatistics.cpp [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+  s.pixelSize = getProperty(Prop::PIXEL_SIZE);
+  s.slit1Slit2Distance = interslitDistance(ws);
+  const std::string slit1SizeEntry = getProperty(Prop::FIRST_SLIT_SIZE_LOG);
+  s.slit1Size = slitSize(ws, slit1SizeEntry);
+  s.slit1SizeDirectBeam = slitSize(directWS, slit1SizeEntry);
+  const std::string slit2Name = getProperty(Prop::SECOND_SLIT_NAME);
+  auto instrument = ws.getInstrument();
+  auto slit2 = instrument->getComponentByName(slit2Name);
+  if (!slit2) {
+    throw Kernel::Exception::NotFoundError("Could not find slit", slit2Name);
```
Comment (reimundILL) [inline Framework/Algorithms/src/ReflectometryBeamStatistics.cpp]: I think you check **several times** if the slit component exists for the same workspace ws (interslitDistance calls slitSeparation and there ...). P.S. Shouldn't it be called interSlitDistance?
I suggest input validation. Then you never need to throw for slits.

File: Framework/Algorithms/src/ReflectometryBeamStatistics.cpp [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+ *
+ * @param ws workspace to investigate
+ * @param logEntry name of the slit opening sample log
+ * @return the slit opening, in meters
+ */
+double ReflectometryBeamStatistics::slitSize(const API::MatrixWorkspace &ws,
+                                             const std::string &logEntry) {
+  auto &run = ws.run();
+  const double opening = run.getPropertyValueAsType<double>(logEntry);
+  const auto &units = run.getProperty(logEntry)->units();
```
Comment (reimundILL) [inline Framework/Algorithms/src/ReflectometryBeamStatistics.cpp]: Very good

File: Framework/Algorithms/test/ReflectometryBeamStatisticsTest.h [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+  static double om_fwhm(const double l2, const double dirl2,
+                        const double dirs2w, const double dirs3w,
+                        const double det_fwhm, const double detdb_fwhm) {
+    // This function comes from COSMOS.
+    using namespace boost::math;
+    const double sdr = SLIT2_DIST + l2;
+    const double ratio = SLIT2_SIZE / SLIT1_SIZE;
+    const double vs = sdr + (ratio * INTERSLIT) / (1 + ratio);
+    const double da_det = std::sqrt(pow<2>(da() * vs) + pow<2>(DET_RESOLUTION));
+    double om_fwhm{0};
```
Comment (reimundILL) [inline Framework/Algorithms/test/ReflectometryBeamStatisticsTest.h]: 0.

File: Framework/Algorithms/test/ReflectometryBeamStatisticsTest.h [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+    if (std::abs(SLIT1_SIZE - dirs2w) >= 0.00004 ||
+        std::abs(SLIT2_SIZE - dirs3w) >= 0.00004) {
+      if ((det_fwhm - da_det) >= 0.) {
+        if (std::sqrt(pow<2>(det_fwhm) - pow<2>(da_det)) >= PIXEL_SIZE) {
+          om_fwhm = 0.5 * std::sqrt(pow<2>(det_fwhm) - pow<2>(da_det)) / dirl2;
+        } else {
+          om_fwhm = 0;
```
Comment (reimundILL) [inline Framework/Algorithms/test/ReflectometryBeamStatisticsTest.h]: 0.

Comment (reimundILL) [reply Framework/Algorithms/test/ReflectometryBeamStatisticsTest.h]: Do you need to explicitly set om_fwhm again to 0.? It is already initialized by 0.

File: Framework/Algorithms/test/ReflectometryBeamStatisticsTest.h [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+  static API::MatrixWorkspace_sptr makeWS(const double braggAngle) {
+    using namespace WorkspaceCreationHelper;
+    using Geometry::deg2rad;
+    constexpr double startX{1000.};
+    const Kernel::V3D sourcePos{0., 0., -L1};
+    const Kernel::V3D &monitorPos = sourcePos;
+    const Kernel::V3D samplePos{
+        0.,
+        0.,
+        0.,
```
Comment (reimundILL) [inline Framework/Algorithms/test/ReflectometryBeamStatisticsTest.h]: Last comma?

File: Framework/Algorithms/src/ReflectometryBeamStatistics.cpp [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+#include "MantidAPI/MatrixWorkspace.h"
+#include "MantidAPI/Run.h"
+#include "MantidAPI/SpectrumInfo.h"
+#include "MantidGeometry/Crystal/AngleUnits.h"
+#include "MantidGeometry/Instrument.h"
+#include "MantidKernel/ArrayLengthValidator.h"
+#include "MantidKernel/BoundedValidator.h"
+#include "MantidKernel/CompositeValidator.h"
+#include "MantidKernel/MandatoryValidator.h"
+#include "MantidKernel/PropertyManagerProperty.h"
```
Comment (reimundILL) [inline Framework/Algorithms/src/ReflectometryBeamStatistics.cpp]: Is `PropertyManagerProperty.h` used?

Comment (reimundILL) [reply Framework/Algorithms/src/ReflectometryBeamStatistics.cpp]: Oh, I see

File: Framework/Algorithms/test/ReflectometryBeamStatisticsTest.h (hunk 5 above)
Comment (soininen) [reply Framework/Algorithms/test/ReflectometryBeamStatisticsTest.h]: It's there for consistency and expandability. What if we ever get a fourth dimension?

Comment (reimundILL) [reply Framework/Algorithms/test/ReflectometryBeamStatisticsTest.h]: For me, V3D is a point with three coordinates (x, y, z) and not a vector.

Comment (reimundILL) [reply Framework/Algorithms/test/ReflectometryBeamStatisticsTest.h]: Thus, I would call it P3D and for a fourth dimension P4D, following the logic.

File: Framework/Algorithms/test/ReflectometryBeamStatisticsTest.h (hunk 4 above)
Comment (soininen) [reply Framework/Algorithms/test/ReflectometryBeamStatisticsTest.h]: The function is straight from COSMOS. In this form it was easier to compare it to the original code for validation but I think we're safe now to remove the quirks.

File: Framework/Algorithms/test/ReflectometryBeamStatisticsTest.h (hunk 5 above)
Comment (reimundILL) [reply Framework/Algorithms/test/ReflectometryBeamStatisticsTest.h]: Seeing V3D as a vector: it's not a general vector

File: Framework/Algorithms/src/ReflectometryBeamStatistics.cpp (hunk 1 above)
Comment (soininen) [reply Framework/Algorithms/src/ReflectometryBeamStatistics.cpp]: I think it should be interslit, see for example international, internet, etc.

**Comment by reimundILL** [comment]
- [x] scripts execute without issues
- [x] sample log values written:

|  | direct | reflected |
|-|--------|------------|
| beam\_rms\_variation | 0.0020868315112967037 | 0.0023237592580850424|
| first\_slit\_angular\_spread | | 7.486762625058678e-05|
| second\_slit\_angular\_spread || 8.39177244498717e-05|
| incident\_angular\_spread || 0.00010926518476230725|
| sample\_waviness || 0.0 |

Question: what about units?

**Comment by reimundILL** [comment]
To be investigated (qthelp):
`WARNING: relatedalgorithms - Could not find algorithm "ReflectometryBeamStatistics, ConvertToReflectometryQ" listed in the seeAlso for ReflectometryMomentumTransfer.v1`

File: Framework/Algorithms/src/ReflectometryMomentumTransfer.cpp [hunk 7]
```diff
@@ -80,7 +96,8 @@ const std::string ReflectometryMomentumTransfer::category() const {
 
 /// Return a vector of related algorithms.
 const std::vector<std::string> ReflectometryMomentumTransfer::seeAlso() const {
-  return {"ConvertToReflectometryQ", "ConvertUnits"};
+  return {"ReflectometryBeamStatistics, ConvertToReflectometryQ",
```
Comment (reimundILL) [inline Framework/Algorithms/src/ReflectometryMomentumTransfer.cpp]: return {"ReflectometryBeamStatistics", "ConvertToReflectometryQ",

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLSumForeground.py [hunk 8]
```diff
@@ -75,17 +70,21 @@ def PyExec(self):
         self._names = common.WSNameSource(wsPrefix, cleanupMode)
 
         ws = self._inputWS()
+        processReflected = not self._directOnly()
+        if processReflected:
+            self._addBeamStatisticsToLogs(ws)
```
Comment (reimundILL) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLSumForeground.py]: Seems to not be documented?

Comment (reimundILL) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLSumForeground.py]: The diagram is updated, but maybe the text can contain some more content or refer to the algorithm?

**Comment by soininen** [comment]
About the units in the sample logs added by `ReflectometryBeamStatistics`:
- `beam_rms_variation`: the unit of pixel size, i.e. meters. The units of intensities seem to cancel out in the calculation which is great since they depend on applied normalizations/calibrations/corrections.
- `first_slit_angular_spread`: radians
- `second_slit_angular_spread`: radians
- `incicent_angular_spread`: radians
- `sample_waviness`: could be radians

**Comment by soininen** [comment]
`ReflectometryILLConvertToQ` documentation should indeed mention that the algorithm need the `foreground.summation_type` sample log. `ReflectometryMomentumTransfer` doesn't use the said log entry.

**Review by vreimund — APPROVED** [review]
Happy about detailed documentation and tests, and overall simplified reduction workflow.
Adding values to the sample logs required by some algorithms remains user friendly due to input validation and the provided documentation. Requires the expert user to better understand the algorithm, this is good and may improve results.
Interfaces break and this is ok since only developers seem to use the underlying reflectometry workflows at the moment. To my best knowledge.
Good work, Antti

**Pull request merged**